lib/hyrax/transactions/steps/save.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+require 'dry/monads'
+
+module Hyrax
+  module Transactions
+    module Steps
+      ##
+      # Saves a given work from a change_set, returning a `Dry::Monads::Result`
+      # (`Success`|`Failure`).
+      #
+      # If the save is successful, publishes an `object.metadata.updated` event
+      # for the affected resource.
+      #
+      # @see https://dry-rb.org/gems/dry-monads/1.0/result/
+      class Save
+        include Dry::Monads[:result]
+
+        ##
+        # @params [#save] persister
+        def initialize(persister: Hyrax.persister)
+          @persister = persister
+        end
+
+        ##
+        # @param [Hyrax::ChangeSet] change_set
+        # @param [::User, nil] user
+        #
+        # @return [Dry::Monads::Result] `Success(work)` if the change_set is
```

Comment:
My reading is that this is correct, we return a `Success(work)` but a `Failure(change_set.resource)`. The reason for the comment is that wrapping different objects in the return may look / feel a bit odd (when it is correct behavior).

_ps_ This is a non-blocking observation

Reply:
`Success` will always wrap the object to be used by the next step; here we transition from a `change_set` to a newly saved resource. 

It's normally the case that `Failure` will wrap something different, whatever is needed to communicate the type of failure and handle it. The convention for what `Failure` wraps is a bit under-determined so far. This one does `['message', change_set.resource]` with the idea that relevant errors from the save will be in the active_model errors. I think i'm on the verge of proposing this as the standard format for errors for Hyrax transactions/steps: `Array<#to_s, Object>` where `#to_s` should be human readable as a description of the error, and `Object` should be something particular to the error type that could aid in reporting, diagnosis, and recovery.